### PR TITLE
authz: Drop requirement for installing authz providers in every service

### PR DESCRIPTION
**Pull request by eseliger**
This is a register call that is easy to forget. When forgotten, all queries against the repo store will block forever.

In addition, this adds a hard-dependency on conf to every services startup, plus a busy loop. With multi-tenant, this will not work great because authz providers would be a global, and we instead want most things to be ephemeral so they're per-provider. This is a step toward that, but doesn't yet remove the providers global variable.

Good news, it turns out that we don't actually need to register the providers in every service! The reason they were required was to check if zero providers are configured, or if authzbypass mode is enabled.

Authz bypass mode is usually ON, except when there are problems with the authz providers, meaning some authz providers might not be able to sync permissions. Bypassing of permissions is only ever happening if there are ALSO zero providers configured.

So this is basically an optimization for the case where an instance has zero authz configured so that the SQL queries are a bit simpler. This also helps in tests because with bypass mode on and no providers configured, authz enforcement is effectively off in the repo store.
This makes it so that in tests we need to do slightly more work, but also makes for a more realistic test vs at runtime setup. Also, it's highly recommended to use mocks for DB wherever possible in more high-level components to keep tests fast.

To never have a scenario where we accidentally mess up here and enable bypass mode erroneously, this PR drops that entirely. Authz is always enforced, but when a code host connection is unrestricted (i.e., will not spawn a provider) the repos are still visible, so this should be no change over before.

## Test plan

The stack starts and works, and all CI tests are still passing. Code review should help as well.

### CODE REVIEW

**Comment by eseliger** [comment]
* **#63743** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/63743?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @eseliger and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: internal/codeintel/syntactic_indexing/jobstore/store.go [hunk 1]
```diff
@@ -169,6 +170,7 @@ SELECT
 	u.should_reindex,
 	u.enqueuer_user_id
 FROM syntactic_scip_indexing_jobs_with_repository_name u
+JOIN repo ON repo.id = u.repository_id AND repo.deleted_at IS NULL AND repo.blocked IS NULL
```
Comment (eseliger) [inline internal/codeintel/syntactic_indexing/jobstore/store.go:173]: this join was missing, the authzquery conds could generate an invalid query here

File: internal/batches/reconciler/executor_test.go [hunk 2]
```diff
@@ -873,7 +873,7 @@ func TestExecutor_ExecutePlan_PublishedChangesetDuplicateBranch(t *testing.T) {
 	}
 
 	logger := logtest.Scoped(t)
-	ctx := context.Background()
+	ctx := actor.WithInternalActor(context.Background())
```
Comment (eseliger) [inline internal/batches/reconciler/executor_test.go:876]: this matches how this is invoked at runtime

File: internal/codeintel/syntactic_indexing/jobstore/store.go (hunk 1 above)
Comment (ggilmore) [reply internal/codeintel/syntactic_indexing/jobstore/store.go:173]: Is this an unrelated fix that you just noticed?

File: internal/authz/providers/bitbucketcloud/authz.go [hunk 3]
```diff
@@ -23,24 +21,8 @@ import (
 // This constructor does not and should not directly check connectivity to external services - if
 // desired, callers should use `(*Provider).ValidateConnection` directly to get warnings related
 // to connection issues.
-func NewAuthzProviders(db database.DB, conns []*types.BitbucketCloudConnection, authProviders []schema.AuthProviders) *atypes.ProviderInitResult {
+func NewAuthzProviders(db database.DB, conns []*types.BitbucketCloudConnection) *atypes.ProviderInitResult {
 	initResults := &atypes.ProviderInitResult{}
-	bbcloudAuthProviders := make(map[string]*schema.BitbucketCloudAuthProvider)
```
Comment (ggilmore) [inline internal/authz/providers/bitbucketcloud/authz.go:28]: I've stared this for a while, but I don't understand quite why this was section was deleted (along with the gerrit one). Can you help me out here?

**Review by ggilmore — COMMENTED** [review]
This looks good overall, but I'm a bit lost on some of the changes (the ones I commented on) - hard for me to tell if those changes are unrelated fixes or of I'm missing something about the impact of the change here

File: internal/codeintel/syntactic_indexing/jobstore/store.go (hunk 1 above)
Comment (eseliger) [reply internal/codeintel/syntactic_indexing/jobstore/store.go:173]: no, now that perms are actually enforced in tests by default this showed in a SQL error during tests 🙈

File: internal/authz/providers/bitbucketcloud/authz.go (hunk 3 above)
Comment (eseliger) [reply internal/authz/providers/bitbucketcloud/authz.go:28]: Maybe I'm totally looking over something here, but I don't think the map that is built here is used anywhere, so this didn't do anything 🤔

**Review by ggilmore — APPROVED** [review]

**Review by pjlast — APPROVED** [review]
Seems good to me. Trusting tests and manual testing here as well

**Pull request merged**